seccomp: Use explicit DefaultErrnoRet

rata opened the pull request:
Since commit "seccomp: Sync fields with runtime-spec fields"
(5d244675bdb23e8fce427036c03517243f344cd4) we support to specify the
DefaultErrnoRet to be used.

Before that commit it was not specified and EPERM was used by default.
This commit keeps the same behaviour but just makes it explicit that the
default is EPERM.

Signed-off-by: Rodrigo Campos <rodrigo@kinvolk.io>

This a follow-up of https://github.com/moby/moby/pull/42604#issuecomment-876632771, as suggested by @thaJeztah. This just adds an explicit default to EPERM. Right now the default is EPERM but is implicit.

<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/moby/moby/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**
Added the field `DefaultErrnoRet` to seccomp profiles

**- How I did it**
Added the field to default profile in `profiles/seccomp/default_linux.go`, then adjusted the tests to expect this new field

**- How to verify it**
You can run unit tests with: `TESTDIRS='github.com/docker/docker/profiles/seccomp' make test-unit`.
You can also run `hack/validate/default-seccomp` to verify nothing was missing in the changes.

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->
Add an explicit DefaultErrnoRet field in the default seccomp profile. No behavior change. 

**- A picture of a cute animal (not mandatory but encouraged)**
![](https://www.kinderundjugendmedien.de/images/Ratatouille_pixar.jpg)

thaJeztah commented on profiles/seccomp/seccomp_test.go:
```diff
@@ -59,26 +61,6 @@ func TestLoadProfile(t *testing.T) {
 	assert.DeepEqual(t, expected, *p)
 }
 
-func TestLoadProfileWithDefaultErrnoRet(t *testing.T) {
```
Wondering why you removed the test here; is it because the one in daemon/ already covers it?

rata replied:
No, because now it is tested in the `TestLoadProfile` function here: https://github.com/moby/moby/pull/42649/files#diff-ee871717077bc31d70d119a348d475d268e45ff60583cc67a294103dd79b11f2R29

I can keep the test if you prefer :)

thaJeztah replied:
Reason I was looking at it was that (but haven't checked all other tests), I felt it was good to keep a test that loads it from the JSON profile (`TestLoadProfile` constructs the config in Go).

rata replied:
Ohh, I see. So, I left this test and also left it `TestLoadProfile`. Let me know if you prefer in some other way :)

thaJeztah commented on daemon/seccomp_linux_test.go:
```diff
@@ -142,7 +145,7 @@ func TestWithSeccomp(t *testing.T) {
 			daemon: &Daemon{
 				seccompEnabled: true,
-				seccompProfile: []byte("{ \"defaultAction\": \"SCMP_ACT_ERRNO\" }"),
+				seccompProfile: []byte("{ \"defaultAction\": \"SCMP_ACT_ERRNO\", \"defaultErrnoRet\": 1 }"),
```
Probably not for this PR, but we should consider using string literals (back ticks) in this test, so that we don't have to do all the escaping for the quotes 😞

rata replied:
heh, that is what I thought when doing the change. I can move it to in this PR or a different one, if you want :)

thaJeztah replied:
Yeah, either way it would cause some code-churn, but I definitely think it makes the test-cases more readable. Perhaps as an extra commit in this PR?

In that case: thinking what would be most useful; perhaps do the backticks change as first commit, so that the "last visible change"  ("git blame") shows the new addition.

(and then 🤞 none of the other reviewers will complain "why make that change?" 😅 🤗)

rata replied:
Will add it as a second commit, in case some reviewer wants to revert it so it is simple. It is slightly more work in case of a git blame, but that is likely to be the case as this mutates too.

Let me know if you think as a first commit is better and I can adapt it :)

thaJeztah reviewed (commented):
Thanks! Note that I also opened https://github.com/moby/moby/pull/42648, which might conflict with this one

rata commented:
@thaJeztah oh, nice PR. I was thinking to open an issue about doing something like that. Feel free to close this and cherry-pick the commit, then! Let me know if you prefer that or if I should polish this PR, Both options are fine for me :)

thaJeztah commented:
If you don't mind the rebase (if my PR gets accepted, and merged first), then I'd say; keep this one open 👍 they're addressing different things, so it makes sense to review/merge separately.

rata commented:
@thaJeztah thanks a lot! I've just rebased on latest master, completed the PR description template and addressed the comments (I think :)).

I've also run `go generate ./profiles/seccomp/` to make sure the generated profile works and `hack/validate/default-seccomp` to validate all is fine :)

Let me know if there is anything else missing!

Thanks again

thaJeztah reviewed (approved):
LGTM, thanks!

thaJeztah commented:
@AkihiroSuda @justincormack PTAL

rata commented:
The CI failure seems unrelated to this PR, IIUC.

justincormack commented on daemon/seccomp_linux_test.go:
```diff
@@ -25,6 +25,8 @@ func TestWithSeccomp(t *testing.T) {
 		comment string
 	}
 
+	defaultErrnoRet := uint(1)
```
can we comment that this is EPERM or use a constant?

rata replied:
Sure, adding a comment!

thaJeztah replied:
should we use syscall.EPERM (or golang.org/x/sys/unix.EPERM) ?

thaJeztah replied:
Looks like https://github.com/moby/moby/pull/42681 used unix.NOSYS, so probably makes sense to do that

rata replied:
Brain fart on my end, this file doesn't need any change at all! All tests make sense without any change, and the test that loads the default profile compares against what `seccomp.GetDefaultProfile()` gives, so it works just fine!

Removing all the changes to `daemon/` files. Sorry! :)

rata replied:
Marking it as resolved as this file is not changed anymore :D

rata committed:
seccomp: Use explicit DefaultErrnoRet

Since commit "seccomp: Sync fields with runtime-spec fields"
(5d244675bdb23e8fce427036c03517243f344cd4) we support to specify the
DefaultErrnoRet to be used.

Before that commit it was not specified and EPERM was used by default.
This commit keeps the same behaviour but just makes it explicit that the
default is EPERM.

Signed-off-by: Rodrigo Campos <rodrigo@kinvolk.io>

rata commented:
@thaJeztah @justincormack PTAL. Now it is only one commit, brain fart on my end, it is not needed to change files in `daemon/` as I explained in this comment: https://github.com/moby/moby/pull/42649#discussion_r679989692

I did add the `defaultErrnoRet` field to the example.json fixture, though, as I want the `TestLoadProfile` to excercise it with that field now that the field is there by default. We do have another function to test this field [here][link], but I think it is worth adding it to the example.json, as that is a more "interesting" example to test the field too.

Please let me know what you think, sorry for the brain fart! :)

[link]: https://github.com/moby/moby/pull/42649/files#diff-ee871717077bc31d70d119a348d475d268e45ff60583cc67a294103dd79b11f2L62-L80

cyphar commented:
So, switching the default to ENOSYS is probably a good thing overall (given that clone3 was missed with the seccomp update it seems the runc behaviour is suboptimal -- though in fairness we are fairly restrained by libseccomp limitations). However, there should be some inverse rules defined to return EPERM for syscalls that are explicitly being blocked (unfortunately due to libseccomp limitations some rules cannot be inverted, but most of the rules Docker uses can be -- I know some podman ones can't though).

rata commented:
@cyphar thanks! The idea is to first add the `DefaultErrnoRet` to the default value (i.e. no behavior change), so it explicit in the profile and users can c&p and easily modify it. Ad then, once this change is done, _discuss_ changing the default or not and how. This is what @thaJeztah suggested [here][link].

Once this is merged I was planning to open an issue to discuss this, I can cc you when creating the issue :) 


[link]: https://github.com/moby/moby/pull/42604#issuecomment-876372997

cpuguy83 commented on profiles/seccomp/default_linux.go:
```diff
@@ -739,9 +739,12 @@ func DefaultProfile() *Seccomp {
 	}
 
+	// The DefaultErrnoRet is 1: EPERM
+	errnoRet := uint(1)
```
Would still be nice to use the constant, then there is no need for a comment.

rata replied:
Using the constant now, thanks!

cyphar commented:
@rata 

> The idea is to first add the DefaultErrnoRet to the default value (i.e. no behavior change), so it is explicit on the profile.

Oh sorry I misread the title of the PR. Yeah that sounds good to me -- sorry for the noise!

cpuguy83 reviewed (approved):
LGTM

thaJeztah reviewed (approved):
LGTM, thanks!
